### PR TITLE
WIP: txnkv/txnsnapshot: Support snapshot set ranges for memory scan

### PR DESCRIPTION
**Pull request by lcwangchao**
Related to issue: https://github.com/pingcap/tidb/issues/26952

Introduce method `SetSortedCustomKeyRetrievers` to read custom ranges. For example, we can do it like below:

```
var snapshot *KVSnapshot
...
retrievers := []*RangeRetriever{NewRangeRetriever(tblPrefix, tblKeyEnd, NewMemDBRetriver(memDB))}
snapshot.SetSortedCustomKeyRetrievers(retrievers)
```

Then the key in range `[tblPrefix, tblKeyEnd)` will be fetched from memDB

### CODE REVIEW

File: tikv/kv.go [hunk 1]
```diff
@@ -582,6 +582,16 @@ type Scanner = txnsnapshot.Scanner
 // KVSnapshot implements the tidbkv.Snapshot interface.
 type KVSnapshot = txnsnapshot.KVSnapshot
 
+// MemDBRetriever implements unionstore.Retriever
+type MemDBRetriever = txnsnapshot.MemDBRetriever
+
+// EmptyRetriever implements unionstore.Retriever
+type EmptyRetriever = txnsnapshot.EmptyRetriever
```
Comment (djshow832) [inline tikv/kv.go:589]: Why do we need `EmptyRetriever`? For the integration of the interfaces?

Comment (lcwangchao) [reply tikv/kv.go:589]: `EmptyRetriever` is used to return an empty data set, i.e, the global temporary table.

**Comment by disksing** [comment]
@lcwangchao Hi, want to continue?

**Comment by lcwangchao** [comment]
> @lcwangchao Hi, want to continue?

Still under review. It is not determined if we should do it yet.

File: tikv/kv.go (hunk 1 above)
Comment (djshow832) [reply tikv/kv.go:589]: I didn't see the usage of `EmptyRetriever` in the other PR. Is it used?

Comment (lcwangchao) [reply tikv/kv.go:589]: It is used in https://github.com/pingcap/tidb/pull/26948 . I think it is also ok to move it to tidb

**Comment by lcwangchao** [comment]
I think it's time to continue

**Pull request closed without merging**